### PR TITLE
Bert qa

### PR DESCRIPTION
**Pull request by spacemanidol**
Adding example for BERT SQUAD with onnx export.

### CODE REVIEW

**Review by bfineran — CHANGES_REQUESTED** [review]
Great job on getting this integration working and in script form.  Really excited about the results, and I had a good time using your previous notebook.

Took a first pass and left some comments.  Didn't comment too much on style because running `make style` will fix a lot of the formatting.  Overall, it would be great if we could structure the readme and script to demonstrate as simple as possible how to integrate sparseML into transformers, pointing this out in both documents.

File: examples/transformers/run_qa.py [hunk 1]
```diff
@@ -0,0 +1,884 @@
...
+        if name_or_regex[:3] == "re:":
+            pattern = name_or_regex[3:]
+            if re.match(pattern, param_name):
+                return True
+        else:
+            if param_name == name_or_regex:
+                return True
+    return False
+
+def get_sparsity_by_regex(
```
[the remark this answers is not in the document]
Comment (spacemanidol) [reply examples/transformers/run_qa.py:455]: I have gone ahead and removed the unused functions.

File: examples/transformers/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# BERT-PRUNE-EXP
+Experiments in pruning BERT
```
[the remark this answers is not in the document]
Comment (spacemanidol) [reply examples/transformers/README.md]: Done.

File: examples/transformers/run_qa.py [hunk 3]
```diff
@@ -0,0 +1,884 @@
...
+        column_names = datasets["train"].column_names
+    else:
+        column_names = datasets["validation"].column_names
+    question_column_name = "question" if "question" in column_names else column_names[0]
+    context_column_name = "context" if "context" in column_names else column_names[1]
+    answer_column_name = "answers" if "answers" in column_names else column_names[2]
+
+    pad_on_right = tokenizer.padding_side == "right"
+
+    def prepare_train_features(examples):
```
[the remark this answers is not in the document]
Comment (spacemanidol) [reply examples/transformers/run_qa.py]: This is how the original scripts from transformers are set up

**Comment by spacemanidol** [comment]
Done with code changes. Working on Readme changes @bfineran  can you give the code another look through for comments/thoughts? README.md will have to wait until all the experiments finish.

File: examples/transformers/README.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+| bert-base-uncased 	|95       	|1                  	|yes      |yes       |0           	|10.501     |4.929      |
+| bert-base-uncased 	|95       	|2                   	|yes      |no        |0            	|24.445     |14.437     |
+| bert-base-uncased 	|95       	|10                 	|yes      |no        |8            	|72.761  	|60.407     |
+| bert-base-uncased 	|99         |1                   	|yes      |yes       |0             |09.685     |03.614     |
+| bert-base-uncased 	|99       	|2                   	|yes      |no        |0            	|17.433     |07.871     |
+| bert-base-uncased 	|99         |10                    	|yes      |no        |8             |47.306    	|32.564     |
+
+
+## Script origin and how to integrate sparseml with other Transformers projects
+This script is based on the example BERT-QA implementation in transformers found [here](https://github.com/huggingface/transformers/blob/master/examples/question-answering/run_qa.py).
```
Comment (markurtz) [inline examples/transformers/README.md:84]: Can we get this to be a direct replacement of the run_qa.py file and have our version be a drop-in file for that?

File: examples/transformers/README.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    get_polynomial_decay_schedule_with_warmup,
+)
+
+from sparseml.pytorch.optim.manager import ScheduledModifierManager
+from sparseml.pytorch.optim.optimizer import ScheduledOptimizer
+from sparseml.pytorch.utils import ModuleExporter
+from sparseml.pytorch.helpers import any_str_or_regex_matches_param_name
+
+
+def get_sparsity_by_regex(module: Module, param_names: List[str]):
```
Comment (markurtz) [inline examples/transformers/README.md:107]: We have functionality similar to this already in sparseML, let's move this function into sparseML

File: examples/transformers/README.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+## Neural Magic Integration here. 
+optim = load_optimizer(model, TrainingArguments) #We first create optimizers based on the method defined in transformers trainer class
+steps_per_epoch = math.ceil(len(datasets["train"]) / (training_args.n_gpu * training_args.per_device_train_batch_size))
+manager = ScheduledModifierManager.from_yaml(data_args.nm_prune_config) # Load a NM pruning config
+
+optim = ScheduledOptimizer(
+    optim, model, manager, steps_per_epoch=steps_per_epoch, loggers=None
+)
+```
+Modify the hugging face trainer to take the sparseml optimzier as shown below
```
Comment (markurtz) [inline examples/transformers/README.md:167]: There's a new diff up from Ben that should make this integration a lot easier by modifying the optimizer in place: https://github.com/neuralmagic/sparseml/pull/77

Can we see if that will simplify this integration so we don't have to pass around our wrapped optimizer?

File: examples/transformers/main.py [hunk 7]
```diff
@@ -0,0 +1,1107 @@
+#!/usr/bin/env python
```
Comment (markurtz) [inline examples/transformers/main.py:1]: Let's rename this file to run_qa.py and ideally have it be a drop in replacement for huggingface's version of that file.

Additionally, we need to add some comments at the top to turn off formatting and copyright for this file. See one of the other integration diffs for an example: https://github.com/neuralmagic/sparseml/pull/70
```
# neuralmagic: no copyright
# flake8: noqa
# fmt: off
# isort: skip_file
```

Let's be sure to call out the portions of the code that we are editing as well with comment blocks, see the above diff for an example, but in general:
    ```
####################################################################################
    # Start SparseML Integration
####################################################################################
```

Finally let's make sure the arguments we're using are consistent with what is being setup in those other integratiions

**Review by markurtz — CHANGES_REQUESTED** [review]
A few things at the high level and left comments:
- let's move this over to a new folder from examples to integrations (the timm and ultrallytics are being migrated there as well)
- let's rename prune_config_files to recipes
-- bonus if we can get these models and recipes pushed to thee sparsezoo instead of storing here
- let's change the formats of the recipes to markdown from yaml and include some info in them such as the example command to run (for when they are pushed to sparsezoo)

File: examples/transformers/README.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+conda create -n sparsemlBERT python=3.7
+conda activate sparsemlBERT
+pip install torch sparseml transformers datasets
+```
+
+## Usage
+To custom prune a model first go to the prune-config.yaml file and modify the parameters to your needs. We have provided a range of pruning configurations in the prune_config_files folder. 
+!EpochRangeModifier controls how long the model trains for and Each !GMPruningModifier modifies controls how each portion is pruned. You can modify end_epoch to control how long the pruning regime lasts and final_sparsity and init_sparsity define the speed which the module is pruned and the final sparsity.
+### Training 
+```python
```
Comment (markurtz) [inline examples/transformers/README.md:19]: format as bash

File: examples/transformers/README.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+ --overwrite_output_dir \
+ --cache_dir cache \
+ --preprocessing_num_workers 4 \
+ --seed 42 \
+ --num_train_epochs 10 \
+ --nm_prune_config prune_config_files/99sparsity10epochs.yaml
+```
+
+#### Evaluation
+```python
```
Comment (markurtz) [inline examples/transformers/README.md:38]: format as bash

File: examples/transformers/README.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+ --dataset_name squad \
+ --do_eval \
+ --per_device_eval_batch_size 12 \
+ --output_dir bert-base-uncased-99sparsity-10total8gmp/ \
+ --overwrite_output_dir \
+ --cache_dir cache \
+ --preprocessing_num_workers 4 \
+```
+#### ONNX Export
+```python
```
Comment (markurtz) [inline examples/transformers/README.md:50]: format as bash

File: examples/transformers/README.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
+# Transformers-SparseML Integration
+This folder contains an example on how to use sparseml with transformers. 
+We focus on Question answering and use a modified implementation from the BERT SQuAD in transformers. 
+Using various pruning configuration files we demostrate the effect unstructured pruning can have on SQuAD.
+
+## Installation and Requirements
+These example scripts require sparseml, transformers, torch, datasets and associated to libraries. To install run the following command
+
+```bash
+conda create -n sparsemlBERT python=3.7
```
Comment (markurtz) [inline examples/transformers/README.md:10]: For the rest of the system, we've been going through instructions with virtualenv's and pip, let's migrate these instructions to that to standardize. No preference on conda over virtualenv for dev's local environments, but need to be consistent in docs

**Review by markurtz — CHANGES_REQUESTED** [review]

File: examples/transformers/README.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+This folder contains an example on how to use sparseml with transformers. 
+We focus on Question answering and use a modified implementation from the BERT SQuAD in transformers. 
+Using various pruning configuration files we demostrate the effect unstructured pruning can have on SQuAD.
+
+## Installation and Requirements
+These example scripts require sparseml, transformers, torch, datasets and associated to libraries. To install run the following command
+
+```bash
+conda create -n sparsemlBERT python=3.8
+conda activate sparsemlBERT
```
[the remark this answers is not in the document]
Comment (spacemanidol) [reply examples/transformers/README.md]: Done. Personally love conda but I get the commonality/consistency component

File: examples/transformers/README.md (hunk 4 above)
Comment (spacemanidol) [reply examples/transformers/README.md:82]: It does that but transformers may update the file. If you look at the bottom of the README.md there are instructions on how to drop into any transforms version

**Comment by spacemanidol** [comment]
Changes implemented. Upgraded to new transformers library and getting a blocking error. Opened an issue on their page. https://github.com/huggingface/transformers/issues/10618

**Comment by spacemanidol** [comment]
Fixed issue with transformers. Ready for merge.

File: examples/transformers/README.md (hunk 6 above)
Comment (spacemanidol) [reply examples/transformers/README.md]: rolled this back because we rolled back this fix.

**Review by markurtz — APPROVED** [review]
As discussed this is in a good state to land for a research integration. Over the next weeks, we need to nail down exact flows and recipes for users (as well as improving the recipes) and then push those to the SparseZoo/potentially huggingface and remove the recipes and table from this integration.

Additionally, as we make things easier we'll need to evaluate how close we can make run_qa.py to the huggingface integration. Layer dropping, distillation, etc should all be added in a separate file.

**Review by bfineran — APPROVED** [review]
excited to see this landed @spacemanidol
